File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    long cardnumber;
    //ask for credit card details and check for a positive number
    do
    {
        printf("Please enter a credit card number: ");
        cardnumber = get_long_long();
    }
    while (cardnumber < 0);

    // check the number of digits

    int count = 0;
    long countdigits = cardnumber;
    while (countdigits > 0)
    {
        countdigits = countdigits / 10;
        count++;
    }
    if ((count != 13) && (count != 15) && (count != 16))
    {
        printf("INVALID\n");
    }
    // create an integer array for storing card numbers
    int number[count];

    for (int i = 0; i < count; i++)
    {
        number[i] = cardnumber % 10;
        cardnumber = cardnumber / 10;
    }
    // create cardnumber in an array
    int cnumber[count];

    for (int i = 0; i < count; i++)
    {
        cnumber[i] = number[i];
    }
    // mulitply every second digit by two
    for (int i = 1; i < count; i += 2)
    {
        number[i] = number[i] * 2;
    }

    int sum = 0;
    int sumdigits;
    // logic for dealing with each card type
    if (count == 13)
    {
        // calculates the sum of each of the card numbers
        for (int i = 0; i < count; i++)
        {
            sumdigits = (number[i] % 10) + (number[i] / 10 % 10);
            sum = sum + sumdigits;
        }
        // determines if Visa card
        if (cnumber[12] == 4 && sum % 10 == 0)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    // logic to check for Amex card based on length
    if (count == 15)
    {
        for (int i = 0; i < count; i++)
        {
            sumdigits = (number[i] % 10) + (number[i] / 10 % 10);
            sum = sum + sumdigits;
        }
        if (cnumber[14] == 3 && sum % 10 == 0 && ((cnumber[13] == 4) || (cnumber[13] == 7)))
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    // logic for MAstercard with 16 digits.  Checks length and first digit.
    if (count == 16 && cnumber[15] == 5)
    {
        for (int i = 0; i < count; i++)
        {
            sumdigits = (number[i] % 10) + (number[i] / 10 % 10);
            sum = sum + sumdigits;
        }
        if (sum % 10 == 0 && ((cnumber[14] == 1) || (cnumber[14] == 2) || (cnumber[14] == 3) || (cnumber[14] == 4) || (cnumber[14] == 5)))
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    // logic for Visa with 16 digits. Checks length and first digit.
    if (count == 16 && cnumber[15] == 4)
    {
        for (int i = 0; i < count; i++)
        {
            sumdigits = (number[i] % 10) + (number[i] / 10 % 10);
            sum = sum + sumdigits;
        }
        if (sum % 10 == 0)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    return 0;
}

File: pset6/credit/credit.py
import cs50
import sys

cardnumber = 0
# ask for credit card details and check for a positive number
while True:
    print("Please enter a credit card number: ")
    cardnumber = cs50.get_int()
    if (cardnumber > 0):
        break

# check the number of digits
if len(str(cardnumber)) != 13 and len(str(cardnumber)) != 15 and len(str(cardnumber)) != 16:
    print("INVALID")

# create an integer array for storing card numbers
number_by_two = []
number = list(map(int, list(str(cardnumber))))
reverse_number = number[::-1]

# mulitply every second digit by two
for i in range(0, len(number)):
    if i % 2 != 0:
        number_by_two.append(reverse_number[i] * 2)
    else:
        number_by_two.append(reverse_number[i])

# get the sum of the card digits with every second number multiplied by 2
card_sum = (sum([x // 10 + x % 10 for x in number_by_two]))

if len(str(cardnumber)) == 13:
    if number[0] == 4 and card_sum % 10 == 0:
        print("VISA")
    else:
        print("INVALID")

# logic to check for Amex card based on length
if len(str(cardnumber)) == 15:
    if number[0] == 3 and card_sum % 10 == 0 and ((number[1] == 4) or (number[1] == 7)):
        print("AMEX")
    else:
        print("INVALID")

# logic for Mastercard with 16 digits.  Checks length and first digit
if len(str(cardnumber)) == 16 and number[0] == 5:
    if card_sum % 10 == 0 and ((number[1] == 1) or (number[1] == 2) or (number[1] == 3) or (number[1] == 4) or (number[1] == 5)):
        print("MASTERCARD")
    else:
        print("INVALID")

# logic for Visa with 16 digits. Checks length and first digit.
if len(str(cardnumber)) == 16 and number[0] == 4:
    if card_sum % 10 == 0:
        print("VISA")
    else:
        print("INVALID")